contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,348 @@
+watches of a given type". As a result, instead of inventing new top level
+pattern per type, we can introduce a new resource type that follows normal
+RESTful rules and solves even more generic problem os spanning multiple
+different resource types.
+
+ We will introduce this new resource in a new dedicated API group:
+```
+/apis/bulk.k8s.io/
+```
+ that underneath will have a separate implementation. Initially we are going
+to support the following operations:
+```
+POST /apis/bulk.k8s.io/v1/get   <body defines filtering>
+```
+ Similarly as in our main api, we will use the same resource for serving both
+LIST and WATCH, and those will differ by appropriately setting `?watch=1 `
+parameter in the request.
+
+ Later, we would like to add ability for bulk create/update/delete with a
+similar pattern, though it's not part of this design.
+
+
+### Filtering definition
+
+ We need to make our filtering mechanism to support different resource types at
+the same time. On the other hand, we would like it to be as consistent with all
+other Kubernetes APIs as possible. So we define the selector for bulk operations
+as following:
+
+```
+// TODO: Find a better name for it?
+type BulkSelector struct {
+	SingleTypeBulkSelector []SingleTypeBulkSelector
+}
+
+// TODO: Find a better name for it?
+type SingleTypeBulkSelector struct {
+	// FIXME: This should be meta.k8s.io/GroupVersionKind instead (or something
+	// similar to it).
+	ResourceType string
+
+	// We would like to reuse the same ListOptions definition as we are using
+	// in regular APIs. For the purpose of bulk operations, you can just specify
+	// more than one.
+	//
+	// NOTE: In the initial implementation, we will require all Options (for a
+	// given resource type) to have exactly the same ResourceVersion - otherwise
+	// we will reject it at the validation level.
+	// This will simplify the initial implementation as we need to guarantee that
+	// watch events within a single resource types are delivered with increasing
+	// resource versions, which may be more tricky with different resource versions.
+	Options []ListOptions
+}
+```
+
+ With such API in place, all List/Watch requests coming from kubelets will be
+allowed to List/Watch only objects that are "attached" to pods bound to a given
+node. We will create a dedicated admission plugin that will be responsible for
+detecting whether a given kubelet (user in general) is allowed to list/watch
+objects requested by `BulkSelector ` (the exact mechanism for it is out of
+scope for this doc) and either rejecting the whole request or letting it go (we
+are not going to support partial rejections (e.g. for this resource you can
+proceed for that one no).
+
+ With the above semantics, whenever user wants to start watching another object
+(or set of objects) or drop one (or a set), he needs to break the watch and
+initiate a new one with different selectors. That isn't perfect, but we still
+can solve all usecases with it.
```

Comment:
But the design as it stands isn't compatible with incrementally changing filters.  I don't see how we extend this design to handle it.